invalid topic / invalid basename errors for certain vehicle classes

While testing #45, I bought some vehicles from the store which apparently have type names which contained some characters "illegal" in a ROS resource name. This resulted in the following warnings:

```
[ERROR] [1631884756.569664824]: isSubtopic() Invalid topic: Character [-] at element [12] is not valid in Graph Resource Name [/ropa_nawaro-maus_111/imu].  Valid characters are a-z, A-Z, 0-9, / and _.
[ERROR] [1631884756.569725723]: isSubtopic() Invalid basename: Character [-] at element [12] is not valid in Graph Resource Name [/ropa_nawaro-maus_111/imu].  Valid characters are a-z, A-Z, 0-9, / and _.
[ERROR] [1631884756.569755232]: isSubtopic() Invalid basename: Character [-] at element [12] is not valid in Graph Resource Name [/ropa_nawaro-maus_111/odom].  Valid characters are a-z, A-Z, 0-9, / and _.
```

seems we need to fix [ros.names.Sanatize(..)](https://github.com/tud-cor/FS19_modROS/blob/c94ccecac93595599943d8faeab74e246933ce04/src/utils/ros.lua#L27-L31).

--- a/src/utils/ros.lua
+++ b/src/utils/ros.lua
@@ -26,8 +26,11 @@ ros.Time = {}
 
 -- function for legalizing a name for ROS resources
 function ros.Names.sanatize(veh_name)
-    -- replace whitespace with underscores and uppercase with lowercase to meet the ROS naming conventions for TF
-    return veh_name:gsub(" ", "_"):lower()
+    -- in order to meet the ROS naming conventions, we need to sanitize illegal names from FS by using
+    -- 1. "%W" to replace every non-alphanumeric character with an underscore "_"
+    -- 2. "+" modifier to get the longest sequence of non-alphanumeric character, i.e. "___" -> "_" (replace consecutive _ with single _)
+    -- 3. :lower() to replace every uppercase letter with lowercase one
+    return veh_name:gsub("%W+", "_"):lower()
 end
 
 

--- a/src/utils/ros.lua
+++ b/src/utils/ros.lua
@@ -30,7 +30,19 @@ function ros.Names.sanatize(veh_name)
     -- 1. "%W" to replace every non-alphanumeric character with an underscore "_"
     -- 2. "+" modifier to get the longest sequence of non-alphanumeric character, i.e. "___" -> "_" (replace consecutive _ with single _)
     -- 3. :lower() to replace every uppercase letter with lowercase one
-    return veh_name:gsub("%W+", "_"):lower()
+    veh_name = veh_name:gsub("%W+", "_"):lower()
+
+    -- make sure names do not start or end with underscores
+    -- if it starts with an underscore, remove the first character
+    if veh_name:sub(1,1) == "_" then
+        veh_name = veh_name:sub(2,-1)
+    end
+
+    -- if it ends with an underscore, remove the last character
+    if veh_name:sub(-1,-1) == "_" then
+        veh_name = veh_name:sub(1,-2)
+    end
+    return veh_name
 end
 
 

--- a/src/utils/ros.lua
+++ b/src/utils/ros.lua
@@ -30,19 +30,15 @@ function ros.Names.sanatize(veh_name)
     -- 1. "%W" to replace every non-alphanumeric character with an underscore "_"
     -- 2. "+" modifier to get the longest sequence of non-alphanumeric character, i.e. "___" -> "_" (replace consecutive _ with single _)
     -- 3. :lower() to replace every uppercase letter with lowercase one
-    veh_name = veh_name:gsub("%W+", "_"):lower()
+    local local_veh_name = veh_name:gsub("%W+", "_"):lower()
 
     -- make sure names do not start or end with underscores
     -- if it starts with an underscore, remove the first character
-    if veh_name:sub(1,1) == "_" then
-        veh_name = veh_name:sub(2,-1)
-    end
-
     -- if it ends with an underscore, remove the last character
-    if veh_name:sub(-1,-1) == "_" then
-        veh_name = veh_name:sub(1,-2)
-    end
-    return veh_name
+    local_veh_name = local_veh_name:gsub("^_", "")
+    local_veh_name = local_veh_name:gsub("_$", "")
+
+    return local_veh_name
 end
 
 

